fabric-registry-sync-v0/src/client/java/net/fabricmc/fabric/impl/client/registry/sync/RegistryRemovalChecker.java

```diff
@@ -0,0 +1,227 @@
+		return missingBlockStates;
+	}
+
+	public static JsonObject serializeRegistries() {
+		JsonObject json = new JsonObject();
+		json.addProperty("version", VERSION);
+
+		// Sort the entries first
+		Map<String, Map<String, Set<String>>> entriesMap = new TreeMap<>();
+		Map<String, Map<String, Set<String>>> blockStatesMap = new TreeMap<>();
+
+		for (Registry<?> registry : Registries.REGISTRIES) {
+			if (registry.size() == 0 || !RegistryAttributeHolder.get(registry).hasAttribute(RegistryAttribute.REMOVAL_CHECKED)) continue;
+
+			Map<String, Set<String>> nsToEntries = new TreeMap<>();
+
+			for (Identifier id : registry.getIds()) {
+				if (!id.getNamespace().equals(Identifier.DEFAULT_NAMESPACE)) {
+					nsToEntries.computeIfAbsent(id.getNamespace(), ns -> new TreeSet<>()).add(id.getPath());
+				}
+			}
+
+			entriesMap.put(registry.getKey().getValue().toString(), nsToEntries);
+		}
+
+		for (Map.Entry<RegistryKey<Block>, Block> blockEntry : Registries.BLOCK.getEntrySet()) {
+			Identifier blockId = blockEntry.getKey().getValue();
+
+			if (
+					blockId.getNamespace().equals(Identifier.DEFAULT_NAMESPACE)
+					|| blockEntry.getValue().getStateManager().getProperties().isEmpty()
+			) {
+				continue;
+			}
+
+			Set<String> states = blockStatesMap.computeIfAbsent(blockId.getNamespace(), ns -> new TreeMap<>()).computeIfAbsent(blockId.getPath(), ns -> new TreeSet<>());
+
+			for (Property<?> property : blockEntry.getValue().getStateManager().getProperties()) {
+				states.add(property.getName());
+			}
+		}
+
+		JsonObject entries = new JsonObject();
+
+		for (Map.Entry<String, Map<String, Set<String>>> registry : entriesMap.entrySet()) {
+			JsonObject registryJson = new JsonObject();
+
+			for (Map.Entry<String, Set<String>> nsToEntries : registry.getValue().entrySet()) {
+				registryJson.add(nsToEntries.getKey(), Util.make(new JsonArray(), arr -> nsToEntries.getValue().forEach(arr::add)));
+			}
+
+			entries.add(registry.getKey(), registryJson);
+		}
+
+		JsonObject blockStates = new JsonObject();
+
+		for (Map.Entry<String, Map<String, Set<String>>> registry : blockStatesMap.entrySet()) {
+			JsonObject registryJson = new JsonObject();
+
+			for (Map.Entry<String, Set<String>> nsToEntries : registry.getValue().entrySet()) {
+				registryJson.add(nsToEntries.getKey(), Util.make(new JsonArray(), arr -> nsToEntries.getValue().forEach(arr::add)));
+			}
+
+			entries.add(registry.getKey(), registryJson);
+		}
+
+		json.add("entries", entries);
+		json.add("blockStates", blockStates);
+		return json;
+	}
+
+	@Nullable
+	public static RegistryRemovalChecker runCheck(Path jsonFile) {
+		if (DISABLED) {
+			LOGGER.info("Registry removal check was disabled via system property.");
+			return null;
+		}
+
+		try {
+			if (Files.exists(jsonFile)) {
+				// Do not run removal check if the file is missing
+				JsonObject json = JsonHelper.deserialize(Files.newBufferedReader(jsonFile, StandardCharsets.UTF_8));
```

Comment:
Thoughts on Gzip'ing the file? Will make it smaller, and also a lot less likely for someone to go and edit it without knowing what it does.